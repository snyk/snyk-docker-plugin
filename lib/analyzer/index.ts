import * as Debug from "debug";
import { Docker, DockerOptions } from "../docker";
import * as dockerFile from "../docker-file";
import * as apkAnalyzer from "./apk-analyzer";
import * as aptAnalyzer from "./apt-analyzer";
import * as binariesAnalyzer from "./binaries-analyzer";
import * as imageInspector from "./image-inspector";
import * as osReleaseDetector from "./os-release-detector";
import * as rpmAnalyzer from "./rpm-analyzer";

export { analyze };

const debug = Debug("snyk");

const extractActions = [
  ...aptAnalyzer.APT_PKGPATHS,
  ...apkAnalyzer.APK_PKGPATHS,
  ...osReleaseDetector.OS_VERPATHS,
].map((p) => {
  return {
    name: "txt",
    pattern: p,
  };
});

async function analyze(
  targetImage: string,
  dockerfileAnalysis?: dockerFile.DockerFileAnalysis,
  options?: DockerOptions,
) {
  const docker = new Docker(targetImage, options);

  await docker.scanStaticalyIfNeeded(extractActions);

  const [imageInspection, osRelease] = await Promise.all([
    imageInspector.detect(docker),
    osReleaseDetector.detect(docker, dockerfileAnalysis),
  ]);

  const results = await Promise.all([
    apkAnalyzer.analyze(docker),
    aptAnalyzer.analyze(docker),
    rpmAnalyzer.analyze(docker),
  ]).catch((err) => {
    debug(`Error while running analyzer: '${err.stderr}'`);
    throw new Error("Failed to detect installed OS packages");
  });

  const { installedPackages, pkgManager } = getInstalledPackages(
    results as any[],
  );
  let binaries;
  try {
    binaries = await binariesAnalyzer.analyze(
      targetImage,
      installedPackages,
      pkgManager,
      options,
    );
  } catch (err) {
    debug(`Error while running binaries analyzer: '${err}'`);
    throw new Error("Failed to detect binaries versions");
  }

  return {
    imageId: imageInspection.Id,
    osRelease,
    results,
    binaries,
    imageLayers: imageInspection.RootFS && imageInspection.RootFS.Layers,
  };
}

function getInstalledPackages(
  results: any[],
): { installedPackages: string[]; pkgManager?: string } {
  const dockerAnalysis = results.find((res) => {
    return res.Analysis && res.Analysis.length > 0;
  });

  if (!dockerAnalysis) {
    return { installedPackages: [] };
  }
  const installedPackages = dockerAnalysis.Analysis.map((pkg) => pkg.Name);
  let pkgManager = dockerAnalysis.AnalyzeType;
  if (pkgManager) {
    pkgManager = pkgManager.toLowerCase();
  }
  return { installedPackages, pkgManager };
}
