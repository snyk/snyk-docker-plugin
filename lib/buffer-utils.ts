import * as crypto from "crypto";
import { HashAlgorithm } from "./types";

const HASH_ENCODING = "hex";

export async function bufferToSha1(buffer: Buffer): Promise<string> {
  const hash = crypto.createHash(HashAlgorithm.Sha1);
  const chunkSize = 100 * 1024 * 1024; // 100 MB

  return new Promise((resolve, reject) => {
    try {
      for (let offset = 0; offset < buffer.length; offset += chunkSize) {
        const end = Math.min(offset + chunkSize, buffer.length);
        const chunk = buffer.slice(offset, end);
        hash.update(chunk);
      }

      const digest = hash.digest(HASH_ENCODING);
      resolve(digest);
    } catch (err) {
      reject(err);
    }
  });
}
