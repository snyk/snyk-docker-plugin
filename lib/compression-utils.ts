import { Decompress as ZstdDecompress } from "fzstd";

/**
 * Decompresses zstd-compressed data from a buffer.
 *
 * This is a synchronous buffer-to-buffer decompression utility.
 * For streaming zstd decompression, use the decompressMaybe transform stream.
 *
 * @param compressed Buffer containing zstd-compressed data
 * @returns Decompressed data as a Buffer
 * @throws Error if decompression fails
 */
export function decompressZstd(compressed: Buffer): Buffer {
  const chunks: Buffer[] = [];

  try {
    const decompressor = new ZstdDecompress((data: Uint8Array) => {
      chunks.push(Buffer.from(data));
    });

    decompressor.push(new Uint8Array(compressed), true);

    return Buffer.concat(chunks);
  } catch (error) {
    throw new Error(
      `Zstd decompression failed: ${
        error instanceof Error ? error.message : String(error)
      }`,
    );
  }
}
