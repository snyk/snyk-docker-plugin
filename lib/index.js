var subProcess = require('./sub-process');
var fetchSnykDockerAnalyzer = require('./fetch-snyk-docker-analyzer');
var debug = require('debug')('snyk');

module.exports = {
  inspect: inspect,
};

function inspect(root, targetFile, options) {
  const targetImage = root;
  return fetchSnykDockerAnalyzer()
  .then(function (analyzerBinaryPath) {
    return Promise.all([
      getRuntime(),
      getDependencies(analyzerBinaryPath, targetImage),
    ])
    .then(function (result) {
      var metadata = {
        name: 'snyk-docker-plugin',
        runtime: result[0],
        packageManager: result[1].packageManager,
      };
      var package = result[1].package;
      return {
        plugin: metadata,
        package: package,
      };
    });
  })
}

function getRuntime() {
  return subProcess.execute('docker', ['version'])
  .then(function (output) {
    versionMatch = /Version:\s+(.*)\n/.exec(output);
    if (versionMatch) {
      return 'docker ' + versionMatch[1]
    }
    return undefined;
  });
}

function handleCommonErrors(error, targetImage) {
  if (error.indexOf('command not found') !== -1) {
    throw new Error('Snyk docker CLI was not found')
  }
  if (error.indexOf('Cannot connect to the Docker daemon') !== -1) {
    throw new Error('Cannot connect to the Docker daemon. Is the docker'
    + ' daemon running?')
  }
  if ((error.indexOf('Error loading image from docker engine') !== -1) ||
      (error.indexOf('Error performing image analysis') !== -1)) {
    throw new Error('Docker image was not found: ' + targetImage)
  }
  if (error.indexOf('Error getting docker client:') !== -1) {
    throw new Error('Failed getting docker client')
  }
  if (error.indexOf('Error processing image:') !== -1) {
    throw new Error('Failed processing image:' + targetImage)
  }
}

function getDependencies(analyzerBinaryPath, targetImage) {
  var result;
  return subProcess.execute(
    analyzerBinaryPath,
    buildArgs(targetImage)
  )
  .then(function (output) {
    result = parseAnalysisResults(output);
    return buildTree(
      targetImage, result.type, result.depInfosList, result.osRelease);
  })
  .then(function (package) {
    return {
      package: package,
      packageManager: result.type,
    };
  })
  .catch(function (error) {
    if (typeof error === 'string') {
      debug(`Error while running analyser: '${error}'`);
      handleCommonErrors(error, targetImage);
      errorMsg = error;
      errorMatch = /msg="(.*)"/g.exec(errorMsg)
      if (errorMatch) {
        errorMsg = errorMatch[1]
      }
      throw new Error(errorMsg);
    }

    throw error;
  })
}

function parseAnalysisResults(analysisOut) {
  var analysisJson = JSON.parse(analysisOut);

  var analysisResult = analysisJson.results.filter(function (res) {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  // TODO: if analysisResult is undefined -
  //  throw a nice error, or report 0 deps

  var depType;
  switch (analysisResult.AnalyzeType) {
    case 'Apt': {
      depType = 'deb';
      break;
    }
    default: {
      depType = analysisResult.AnalyzeType.toLowerCase()
    }
  }

  return {
    type: depType,
    depInfosList: analysisResult.Analysis,
    osRelease: analysisJson.osRelease,
  }
}

function buildTree(targetImage, depType, depInfosList, osRelease) {
  var targetSplit = targetImage.split(':');
  var imageName = targetSplit[0];
  var imageVersion = targetSplit[1] ? targetSplit[1] : 'latest';

  var root = {
    name: imageName,
    version: imageVersion,
    dockerOSRelease: osRelease,
    packageFormatVersion: depType + ':0.0.1',
    dependencies: {},
  };

  var depsMap = depInfosList.reduce(function (acc, depInfo) {
    var name = depInfo.Name;
    acc[name] = depInfo;
    return acc;
  }, {});

  var virtualDepsMap = depInfosList.reduce(function (acc, depInfo) {
    var providesNames = depInfo.Provides || [];
    providesNames.forEach(function (name) {
      acc[name] = depInfo;
    });
    return acc;
  }, {});

  var attachDeps = function (depInfos) {
    depInfos.forEach(function (depInfo) {
      var subtree = buildTreeRecurisve(
        depInfo.Name, new Set(), depsMap, virtualDepsMap);
      if (subtree) {
        root.dependencies[subtree.name] = subtree;
      }
    })
  };

  // attach (as direct deps) pkgs not marked auto-installed:
  var manuallyInstalledDeps = depInfosList.filter(function (depInfo) {
    return !depInfo.AutoInstalled;
  })
  attachDeps(manuallyInstalledDeps);

  // attach (as direct deps) pkgs marked as auto-insatalled,
  //  but not dependant upon:
  var notVisitedDeps = depInfosList.filter(function (depInfo) {
    var depName = depInfo.Name;
    return !(depsMap[depName]._visited);
  })
  attachDeps(notVisitedDeps);

  return root;
}

function buildTreeRecurisve(
    depName, ancestors, depsMap, virtualDepsMap) {
  var depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return null;
  }

  var fullName = depFullName(depInfo);
  if (ancestors.has(fullName)) {
    return null;
  }

  const tree = {
    name: fullName,
    version: depInfo.Version,
  };

  //TODO: if we do this we can avoid tracking ancestors
  if (depInfo._visited) {
    return tree;
  }
  depInfo._visited = true;

  var newAncestors = (new Set(ancestors)).add(fullName);

  var deps = depInfo.Deps || {};
  Object.keys(deps).forEach(function (depName) {
    var subTree = buildTreeRecurisve(
      depName, newAncestors, depsMap, virtualDepsMap);
    if (subTree) {
      if (!tree.dependencies) {
        tree.dependencies = {};
      }
      tree.dependencies[subTree.name] = subTree;
    }
  });

  return tree;
}

function depFullName(depInfo) {
  var fullName = depInfo.Name;
  if (depInfo.Source) {
    fullName = depInfo.Source + '/' + fullName;
  }
  return fullName;
}

function buildArgs(targetImage) {
  var args = ['analyze', targetImage];
  return args;
}
