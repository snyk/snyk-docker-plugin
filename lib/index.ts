const debug = require('debug')('snyk');

import * as analyzer from './analyzer';
import * as subProcess from './sub-process';
import * as dockerFile from './docker-file';

export {
  inspect,
};

function inspect(root: string, targetFile?: string, options?: any) {
  const targetImage = root;
  return Promise.all([
    getRuntime(),
    getDependencies(targetImage),
    dockerFile.getBaseImageName(targetFile),
  ])
    .then((result) => {
      const metadata = {
        name: 'snyk-docker-plugin',
        runtime: result[0],
        packageManager: result[1].packageManager,
        dockerImageId: result[1].imageId,
      };
      const pkg: any = result[1].package;
      pkg.docker = pkg.docker || {};
      pkg.docker.baseImage = result[2];
      return {
        plugin: metadata,
        package: pkg,
      };
    });
}

function getRuntime() {
  return subProcess.execute('docker', ['version'])
    .then((output) => {
      const versionMatch = /Version:\s+(.*)\n/.exec(output);
      if (versionMatch) {
        return 'docker ' + versionMatch[1];
      }
      return undefined;
    });
}

function handleCommonErrors(error, targetImage: string) {
  if (error.indexOf('command not found') !== -1) {
    throw new Error('Snyk docker CLI was not found');
  }
  if (error.indexOf('Cannot connect to the Docker daemon') !== -1) {
    throw new Error('Cannot connect to the Docker daemon. Is the docker'
    + ' daemon running?');
  }
  const ERROR_LOADING_IMAGE_STR = 'Error loading image from docker engine:';
  if (error.indexOf(ERROR_LOADING_IMAGE_STR) !== -1) {
    if (error.indexOf('reference does not exist') !== -1) {
      throw new Error(
        `Docker image was not found locally: ${targetImage}`);
    }
    if (error.indexOf('permission denied while trying to connect') !== -1) {
      let errString = error.split(ERROR_LOADING_IMAGE_STR)[1];
      errString = (errString || '').slice(0, -2); // remove trailing \"
      throw new Error(
        'Permission denied connecting to docker daemon. ' +
        'Please make sure user has the required permissions. ' +
        'Error string: ' + errString);
    }
  }
  if (error.indexOf('Error getting docker client:') !== -1) {
    throw new Error('Failed getting docker client');
  }
  if (error.indexOf('Error processing image:') !== -1) {
    throw new Error('Failed processing image:' + targetImage);
  }
}

function getDependencies(targetImage: string) {
  let result;
  return analyzer.analyze(targetImage)
    .then((output) => {
      result = parseAnalysisResults(output);
      return buildTree(
        targetImage, result.type, result.depInfosList, result.targetOS);
    })
    .then((pkg) => {
      return {
        package: pkg,
        packageManager: result.type,
        imageId: result.imageId,
      };
    })
    .catch((error) => {
      if (typeof error === 'string') {
        debug(`Error while running analyzer: '${error}'`);
        handleCommonErrors(error, targetImage);
        let errorMsg = error;
        const errorMatch = /msg="(.*)"/g.exec(errorMsg);
        if (errorMatch) {
          errorMsg = errorMatch[1];
        }
        throw new Error(errorMsg);
      }
      throw error;
    });
}

function parseAnalysisResults(analysisJson) {
  const analysisResult = analysisJson.results.filter((res) => {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  if (!analysisResult) {
    throw new Error(
      'Failed to detect a supported Linux package manager (deb/rpm/apk)');
  }

  let depType;
  switch (analysisResult.AnalyzeType) {
    case 'Apt': {
      depType = 'deb';
      break;
    }
    default: {
      depType = analysisResult.AnalyzeType.toLowerCase();
    }
  }

  return {
    imageId: analysisJson.imageId,
    targetOS: analysisJson.osRelease,
    type: depType,
    depInfosList: analysisResult.Analysis,
  };
}

function buildTree(targetImage, depType, depInfosList, targetOS) {
  const targetSplit = targetImage.split(':');
  const imageName = targetSplit[0];
  const imageVersion = targetSplit[1] ? targetSplit[1] : 'latest';

  const root = {
    name: imageName,
    version: imageVersion,
    targetOS,
    packageFormatVersion: depType + ':0.0.1',
    dependencies: {},
  };

  const depsMap = depInfosList.reduce((acc, depInfo) => {
    const name = depInfo.Name;
    acc[name] = depInfo;
    return acc;
  }, {});

  const virtualDepsMap = depInfosList.reduce((acc, depInfo) => {
    const providesNames = depInfo.Provides || [];
    providesNames.forEach((name) => {
      acc[name] = depInfo;
    });
    return acc;
  }, {});

  const depsCounts = {};
  depInfosList.forEach((depInfo) => {
    countDepsRecursive(
      depInfo.Name, new Set(), depsMap, virtualDepsMap, depsCounts);
  });
  const DEP_FREQ_THRESHOLD = 100;
  const tooFrequentDepNames = Object.keys(depsCounts)
    .filter((depName) => {
      return depsCounts[depName] > DEP_FREQ_THRESHOLD;
    });

  const attachDeps = (depInfos) => {
    const depNamesToSkip = new Set(tooFrequentDepNames);
    depInfos.forEach((depInfo) => {
      const subtree = buildTreeRecurisve(
        depInfo.Name, new Set(), depsMap, virtualDepsMap, depNamesToSkip);
      if (subtree) {
        root.dependencies[subtree.name] = subtree;
      }
    });
  };

  // attach (as direct deps) pkgs not marked auto-installed:
  const manuallyInstalledDeps = depInfosList.filter((depInfo) => {
    return !depInfo.AutoInstalled;
  });
  attachDeps(manuallyInstalledDeps);

  // attach (as direct deps) pkgs marked as auto-insatalled,
  //  but not dependant upon:
  const notVisitedDeps = depInfosList.filter((depInfo) => {
    const depName = depInfo.Name;
    return !(depsMap[depName]._visited);
  });
  attachDeps(notVisitedDeps);

  // group all the "too frequest" deps under a meta package:
  if (tooFrequentDepNames.length > 0) {
    const tooFrequentDeps = tooFrequentDepNames.map((name) => {
      return depsMap[name];
    });

    const metaSubtree = {
      name: 'meta-common-packages',
      version: 'meta',
      dependencies: {},
    };

    tooFrequentDeps.forEach((depInfo) => {
      const pkg = {
        name: depFullName(depInfo),
        version: depInfo.Version,
      };
      metaSubtree.dependencies[pkg.name] = pkg;
    });

    root.dependencies[metaSubtree.name] = metaSubtree;
  }

  return root;
}

function buildTreeRecurisve(
  depName, ancestors, depsMap, virtualDepsMap, depNamesToSkip) {
  const depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return null;
  }

  // "realName" as the argument depName might be a virtual pkg
  const realName = depInfo.Name;
  const fullName = depFullName(depInfo);
  if (ancestors.has(fullName) || depNamesToSkip.has(realName)) {
    return null;
  }

  const tree: {
    name: string;
    version: string;
    dependencies?: any;
  } = {
    name: fullName,
    version: depInfo.Version,
  };

  if (depInfo._visited) {
    return tree;
  }
  depInfo._visited = true;

  const newAncestors = (new Set(ancestors)).add(fullName);

  const deps = depInfo.Deps || {};
  Object.keys(deps).forEach((name) => {
    const subTree = buildTreeRecurisve(
      name, newAncestors, depsMap, virtualDepsMap, depNamesToSkip);
    if (subTree) {
      if (!tree.dependencies) {
        tree.dependencies = {};
      }
      tree.dependencies[subTree.name] = subTree;
    }
  });

  return tree;
}

function countDepsRecursive(
  depName, ancestors, depsMap, virtualDepsMap, depCounts) {
  const depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return;
  }

  // "realName" as the argument depName might be a virtual pkg
  const realName = depInfo.Name;
  if (ancestors.has(realName)) {
    return;
  }

  depCounts[realName] = (depCounts[realName] || 0) + 1;

  const newAncestors = (new Set(ancestors)).add(realName);
  const deps = depInfo.Deps || {};
  Object.keys(deps).forEach((name) => {
    countDepsRecursive(
      name, newAncestors, depsMap, virtualDepsMap, depCounts);
  });
}

function depFullName(depInfo) {
  let fullName = depInfo.Name;
  if (depInfo.Source) {
    fullName = depInfo.Source + '/' + fullName;
  }
  return fullName;
}
