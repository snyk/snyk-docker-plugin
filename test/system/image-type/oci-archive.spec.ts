import { scan } from "../../../lib";
import { getFixture } from "../../util";

describe("oci archive scanning", () => {
  it("should correctly scan an oci archive", async () => {
    const fixturePath = getFixture("oci-archives/alpine-3.12.0.tar");
    const imageNameAndTag = `oci-archive:${fixturePath}`;

    const pluginResult = await scan({
      path: imageNameAndTag,
    });

    expect(pluginResult).toMatchSnapshot();
  });

  it("should correctly scan a busybox oci archive", async () => {
    const fixturePath = getFixture("oci-archives/busybox-1.31.1.tar");
    const imageNameAndTag = `oci-archive:${fixturePath}`;

    const pluginResult = await scan({
      path: imageNameAndTag,
    });
    expect(pluginResult).toMatchSnapshot();
  });
});

describe("handles bad input being provided", () => {
  it("should reject when provided with a non-existent oci-archive", async () => {
    await expect(() =>
      scan({
        path: "oci-archive:not-here.tar",
      }),
    ).rejects.toEqual(
      Error("The provided archive path does not exist on the filesystem"),
    );
  });
});
