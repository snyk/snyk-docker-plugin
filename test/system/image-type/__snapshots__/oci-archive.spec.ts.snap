// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oci archive scanning should correctly scan a busybox oci archive 1`] = `
Object {
  "scanResults": Array [
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "graph": Object {
              "nodes": Array [
                Object {
                  "deps": Array [],
                  "nodeId": "root-node",
                  "pkgId": "docker-image|busybox-1.31.1.tar@",
                },
              ],
              "rootNodeId": "root-node",
            },
            "pkgManager": Object {
              "name": "linux",
              "repositories": Array [
                Object {
                  "alias": "unknown:0.0",
                },
              ],
            },
            "pkgs": Array [
              Object {
                "id": "docker-image|busybox-1.31.1.tar@",
                "info": Object {
                  "name": "docker-image|busybox-1.31.1.tar",
                  "version": undefined,
                },
              },
            ],
            "schemaVersion": "1.2.0",
          },
          "type": "depGraph",
        },
        Object {
          "data": "sha256:2f2d4db2ef818c1c897eb71421e8a250daf19e21a4ec4c00a549b16ed0c2b46b",
          "type": "imageId",
        },
        Object {
          "data": Array [
            "sha256:76df9210b28cbd4bc127844914d0a23937ed213048dc6289b2a2d4f7d675c75e",
          ],
          "type": "imageLayers",
        },
        Object {
          "data": Array [
            "sha256:1be74353c3d0fd55fb5638a52953e6f1bc441e5b1710921db9ec2aa202725569",
          ],
          "type": "rootFs",
        },
      ],
      "identity": Object {
        "args": Object {
          "platform": "linux/amd64",
        },
        "type": "linux",
      },
      "target": Object {
        "image": "docker-image|busybox-1.31.1.tar",
      },
    },
  ],
}
`;

exports[`oci archive scanning should correctly scan an oci archive 1`] = `
Object {
  "scanResults": Array [
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "graph": Object {
              "nodes": Array [
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "alpine-baselayout/alpine-baselayout@3.2.0-r6",
                    },
                    Object {
                      "nodeId": "alpine-keys/alpine-keys@2.2-r0",
                    },
                    Object {
                      "nodeId": "apk-tools/apk-tools@2.10.5-r1",
                    },
                    Object {
                      "nodeId": "busybox/busybox@1.31.1-r16|2",
                    },
                    Object {
                      "nodeId": "busybox/ssl_client@1.31.1-r16",
                    },
                    Object {
                      "nodeId": "ca-certificates/ca-certificates-bundle@20191127-r2",
                    },
                    Object {
                      "nodeId": "libc-dev/libc-utils@0.7.2-r3",
                    },
                    Object {
                      "nodeId": "libtls-standalone/libtls-standalone@2.9.1-r1|2",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                    Object {
                      "nodeId": "musl/musl-utils@1.1.24-r8|2",
                    },
                    Object {
                      "nodeId": "openssl/libcrypto1.1@1.1.1g-r0|2",
                    },
                    Object {
                      "nodeId": "openssl/libssl1.1@1.1.1g-r0|2",
                    },
                    Object {
                      "nodeId": "pax-utils/scanelf@1.2.6-r0|2",
                    },
                    Object {
                      "nodeId": "zlib/zlib@1.2.11-r3|2",
                    },
                  ],
                  "nodeId": "root-node",
                  "pkgId": "docker-image|alpine-3.12.0.tar@",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "busybox/busybox@1.31.1-r16|1",
                  "pkgId": "busybox/busybox@1.31.1-r16",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                  ],
                  "nodeId": "busybox/busybox@1.31.1-r16|2",
                  "pkgId": "busybox/busybox@1.31.1-r16",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "musl/musl@1.1.24-r8",
                  "pkgId": "musl/musl@1.1.24-r8",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "busybox/busybox@1.31.1-r16|1",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                  ],
                  "nodeId": "alpine-baselayout/alpine-baselayout@3.2.0-r6",
                  "pkgId": "alpine-baselayout/alpine-baselayout@3.2.0-r6",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "alpine-keys/alpine-keys@2.2-r0",
                  "pkgId": "alpine-keys/alpine-keys@2.2-r0",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "openssl/libcrypto1.1@1.1.1g-r0|1",
                  "pkgId": "openssl/libcrypto1.1@1.1.1g-r0",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                  ],
                  "nodeId": "openssl/libcrypto1.1@1.1.1g-r0|2",
                  "pkgId": "openssl/libcrypto1.1@1.1.1g-r0",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "openssl/libssl1.1@1.1.1g-r0|1",
                  "pkgId": "openssl/libssl1.1@1.1.1g-r0",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                    Object {
                      "nodeId": "openssl/libcrypto1.1@1.1.1g-r0|1",
                    },
                  ],
                  "nodeId": "openssl/libssl1.1@1.1.1g-r0|2",
                  "pkgId": "openssl/libssl1.1@1.1.1g-r0",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "zlib/zlib@1.2.11-r3|1",
                  "pkgId": "zlib/zlib@1.2.11-r3",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                  ],
                  "nodeId": "zlib/zlib@1.2.11-r3|2",
                  "pkgId": "zlib/zlib@1.2.11-r3",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                    Object {
                      "nodeId": "openssl/libcrypto1.1@1.1.1g-r0|1",
                    },
                    Object {
                      "nodeId": "openssl/libssl1.1@1.1.1g-r0|1",
                    },
                    Object {
                      "nodeId": "zlib/zlib@1.2.11-r3|1",
                    },
                  ],
                  "nodeId": "apk-tools/apk-tools@2.10.5-r1",
                  "pkgId": "apk-tools/apk-tools@2.10.5-r1",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "libtls-standalone/libtls-standalone@2.9.1-r1|1",
                  "pkgId": "libtls-standalone/libtls-standalone@2.9.1-r1",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "ca-certificates/ca-certificates-bundle@20191127-r2",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                    Object {
                      "nodeId": "openssl/libcrypto1.1@1.1.1g-r0|1",
                    },
                    Object {
                      "nodeId": "openssl/libssl1.1@1.1.1g-r0|1",
                    },
                  ],
                  "nodeId": "libtls-standalone/libtls-standalone@2.9.1-r1|2",
                  "pkgId": "libtls-standalone/libtls-standalone@2.9.1-r1",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "libtls-standalone/libtls-standalone@2.9.1-r1|1",
                    },
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                  ],
                  "nodeId": "busybox/ssl_client@1.31.1-r16",
                  "pkgId": "busybox/ssl_client@1.31.1-r16",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "ca-certificates/ca-certificates-bundle@20191127-r2",
                  "pkgId": "ca-certificates/ca-certificates-bundle@20191127-r2",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "musl/musl-utils@1.1.24-r8|1",
                  "pkgId": "musl/musl-utils@1.1.24-r8",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                    Object {
                      "nodeId": "pax-utils/scanelf@1.2.6-r0|1",
                    },
                  ],
                  "nodeId": "musl/musl-utils@1.1.24-r8|2",
                  "pkgId": "musl/musl-utils@1.1.24-r8",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl-utils@1.1.24-r8|1",
                    },
                  ],
                  "nodeId": "libc-dev/libc-utils@0.7.2-r3",
                  "pkgId": "libc-dev/libc-utils@0.7.2-r3",
                },
                Object {
                  "deps": Array [],
                  "nodeId": "pax-utils/scanelf@1.2.6-r0|1",
                  "pkgId": "pax-utils/scanelf@1.2.6-r0",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "musl/musl@1.1.24-r8",
                    },
                  ],
                  "nodeId": "pax-utils/scanelf@1.2.6-r0|2",
                  "pkgId": "pax-utils/scanelf@1.2.6-r0",
                },
              ],
              "rootNodeId": "root-node",
            },
            "pkgManager": Object {
              "name": "apk",
              "repositories": Array [
                Object {
                  "alias": "alpine:3.12.0",
                },
              ],
            },
            "pkgs": Array [
              Object {
                "id": "docker-image|alpine-3.12.0.tar@",
                "info": Object {
                  "name": "docker-image|alpine-3.12.0.tar",
                  "version": undefined,
                },
              },
              Object {
                "id": "busybox/busybox@1.31.1-r16",
                "info": Object {
                  "name": "busybox/busybox",
                  "version": "1.31.1-r16",
                },
              },
              Object {
                "id": "musl/musl@1.1.24-r8",
                "info": Object {
                  "name": "musl/musl",
                  "version": "1.1.24-r8",
                },
              },
              Object {
                "id": "alpine-baselayout/alpine-baselayout@3.2.0-r6",
                "info": Object {
                  "name": "alpine-baselayout/alpine-baselayout",
                  "version": "3.2.0-r6",
                },
              },
              Object {
                "id": "alpine-keys/alpine-keys@2.2-r0",
                "info": Object {
                  "name": "alpine-keys/alpine-keys",
                  "version": "2.2-r0",
                },
              },
              Object {
                "id": "openssl/libcrypto1.1@1.1.1g-r0",
                "info": Object {
                  "name": "openssl/libcrypto1.1",
                  "version": "1.1.1g-r0",
                },
              },
              Object {
                "id": "openssl/libssl1.1@1.1.1g-r0",
                "info": Object {
                  "name": "openssl/libssl1.1",
                  "version": "1.1.1g-r0",
                },
              },
              Object {
                "id": "zlib/zlib@1.2.11-r3",
                "info": Object {
                  "name": "zlib/zlib",
                  "version": "1.2.11-r3",
                },
              },
              Object {
                "id": "apk-tools/apk-tools@2.10.5-r1",
                "info": Object {
                  "name": "apk-tools/apk-tools",
                  "version": "2.10.5-r1",
                },
              },
              Object {
                "id": "libtls-standalone/libtls-standalone@2.9.1-r1",
                "info": Object {
                  "name": "libtls-standalone/libtls-standalone",
                  "version": "2.9.1-r1",
                },
              },
              Object {
                "id": "busybox/ssl_client@1.31.1-r16",
                "info": Object {
                  "name": "busybox/ssl_client",
                  "version": "1.31.1-r16",
                },
              },
              Object {
                "id": "ca-certificates/ca-certificates-bundle@20191127-r2",
                "info": Object {
                  "name": "ca-certificates/ca-certificates-bundle",
                  "version": "20191127-r2",
                },
              },
              Object {
                "id": "musl/musl-utils@1.1.24-r8",
                "info": Object {
                  "name": "musl/musl-utils",
                  "version": "1.1.24-r8",
                },
              },
              Object {
                "id": "libc-dev/libc-utils@0.7.2-r3",
                "info": Object {
                  "name": "libc-dev/libc-utils",
                  "version": "0.7.2-r3",
                },
              },
              Object {
                "id": "pax-utils/scanelf@1.2.6-r0",
                "info": Object {
                  "name": "pax-utils/scanelf",
                  "version": "1.2.6-r0",
                },
              },
            ],
            "schemaVersion": "1.2.0",
          },
          "type": "depGraph",
        },
        Object {
          "data": "sha256:0f5f445df8ccbd8a062ad3d02d459e8549d9998c62a5b7cbf77baf68aa73bf5b",
          "type": "imageId",
        },
        Object {
          "data": Array [
            "sha256:df20fa9351a15782c64e6dddb2d4a6f50bf6d3688060a34c4014b0d9a752eb4c",
          ],
          "type": "imageLayers",
        },
        Object {
          "data": Array [
            "sha256:50644c29ef5a27c9a40c393a73ece2479de78325cae7d762ef3cdc19bf42dd0a",
          ],
          "type": "rootFs",
        },
        Object {
          "data": "Alpine Linux v3.12",
          "type": "imageOsReleasePrettyName",
        },
      ],
      "identity": Object {
        "args": Object {
          "platform": "linux/amd64",
        },
        "type": "apk",
      },
      "target": Object {
        "image": "docker-image|alpine-3.12.0.tar",
      },
    },
  ],
}
`;
